chunk/schema.go

```diff
@@ -618,128 +464,105 @@ func (v6Entries) GetWriteEntries(_, through uint32, tableName, hashKey string, l
 			Value:      []byte(value),
 		})
 	}
 
 	return entries, nil
 }
 
-func (v6Entries) GetReadMetricEntries(from, _ uint32, tableName, hashKey string) ([]IndexEntry, error) {
+func (v6Entries) GetReadQueries(_, _ uint32, _, _ string) ([]IndexQuery, error) {
+	return nil, fmt.Errorf("v6Entries does not support GetReadQueries")
+}
+
+func (v6Entries) GetReadMetricQueries(from, _ uint32, tableName, bucketHashKey string, metricName model.LabelValue) ([]IndexQuery, error) {
 	encodedFromBytes := encodeTime(from)
-	return []IndexEntry{
+	return []IndexQuery{
 		{
 			TableName:       tableName,
-			HashValue:       hashKey,
+			HashValue:       bucketHashKey + ":" + string(metricName),
 			RangeValueStart: buildRangeKey(encodedFromBytes),
 		},
 	}, nil
 }
 
-func (v6Entries) GetReadMetricLabelEntries(from, _ uint32, tableName, hashKey string, labelName model.LabelName) ([]IndexEntry, error) {
+func (v6Entries) GetReadMetricLabelQueries(from, _ uint32, tableName, bucketHashKey string, metricName model.LabelValue, labelName model.LabelName) ([]IndexQuery, error) {
 	encodedFromBytes := encodeTime(from)
-	return []IndexEntry{
+	return []IndexQuery{
 		{
 			TableName:       tableName,
-			HashValue:       hashKey + ":" + string(labelName),
+			HashValue:       bucketHashKey + ":" + string(metricName) + ":" + string(labelName),
 			RangeValueStart: buildRangeKey(encodedFromBytes),
 		},
 	}, nil
 }
 
-func (v6Entries) GetReadMetricLabelValueEntries(from, _ uint32, tableName, hashKey string, labelName model.LabelName, labelValue model.LabelValue) ([]IndexEntry, error) {
+func (v6Entries) GetReadMetricLabelValueQueries(from, _ uint32, tableName, bucketHashKey string, metricName model.LabelValue, labelName model.LabelName, labelValue model.LabelValue) ([]IndexQuery, error) {
 	encodedFromBytes := encodeTime(from)
-	return []IndexEntry{
+	return []IndexQuery{
 		{
 			TableName:       tableName,
-			HashValue:       hashKey + ":" + string(labelName),
+			HashValue:       bucketHashKey + ":" + string(metricName) + ":" + string(labelName),
 			RangeValueStart: buildRangeKey(encodedFromBytes),
 		},
 	}, nil
 }
 
-func buildRangeKey(ss ...[]byte) []byte {
-	length := 0
-	for _, s := range ss {
-		length += len(s) + 1
-	}
-	output, i := make([]byte, length, length), 0
-	for _, s := range ss {
-		copy(output[i:i+len(s)], s)
-		i += len(s) + 1
-	}
-	return output
+// v7Entries supports queries with no metric name
+type v7Entries struct {
+	v6Entries
 }
 
-func encodeBase64Value(value model.LabelValue) []byte {
-	encodedLen := base64.RawStdEncoding.EncodedLen(len(value))
-	encoded := make([]byte, encodedLen, encodedLen)
-	base64.RawStdEncoding.Encode(encoded, []byte(value))
-	return encoded
-}
+func (v7Entries) GetWriteEntries(_, through uint32, tableName, bucketHashKey string, metricName model.LabelValue, labels model.Metric, chunkID string) ([]IndexEntry, error) {
+	metricName, err := util.ExtractMetricNameFromMetric(labels)
```

Comment:
Please call `entries.v6Entries.GetWriteEntries(...)` then add the extra entry to that, should reduce code duplication.